atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java

```diff
@@ -152,14 +152,29 @@
         implements AsyncInitializer, CassandraKeyValueService {
 
     public static class InitializeCheckingWrapper implements AutoDelegate_CassandraKeyValueService {
+        private enum Status {
+            OPEN, CLOSING, CLOSED
+        }
+
         private CassandraKeyValueServiceImpl kvs;
+        private volatile Status status = Status.OPEN;
 
         public InitializeCheckingWrapper(CassandraKeyValueServiceImpl kvs) {
             this.kvs = kvs;
         }
 
         @Override
         public CassandraKeyValueService delegate() {
+            if (status == Status.CLOSING) {
+                close();
+                if (status != Status.CLOSED) {
+                    throw new NotInitializedException("CassandraKeyValueService");
+                }
```

Comment:
Should we return the delegate even if we just successfully closed?
hat-tip @fsamuel-bs

Reply:
Maybe throw to match the standard behaviour of `Closeable` things?